src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs

```diff
@@ -521,6 +521,12 @@ public class Options
             [Argument(ArgumentType.AtMostOnce, HelpText = "Maximum number of rows to produce", ShortName = "rows", Hide = true)]
             public long? MaxRows;
 
+            /// <summary>
+            /// Character to use to escape quotes inside quoted fields.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Character to use to escape quotes inside quoted fields", ShortName = "escapechar")]
+            public char EscapeChar = Defaults.EscapeChar;
+
```

Comment:
_(pinned to a line of code to allow for threaded conversations)_

So currently this PR uses the escapeChar to escape quote chars (") inside quoted fields; which I think should be enough for the main ask made by Justin of letting users escape with `\"`. Still, there is room for expanding this feature in 2 directions:

1. **Allow escapeChar to also escape quote chars inside non-quoted fields**. The [RFC 4180](https://tools.ietf.org/html/rfc4180) would actually consider invalid any quote char (") outside of quoted fields.  Still, ML.NET has always allowed unescaped quotes inside non-quoted fields... so if a field contained something like `1,these are 2 quotes "" and this is one " quote,true` it would actually read the second field as `these are 2 quotes "" and this is one " quote` (i.e., without throwing or logging errors, and without escaping the first ""). So if we allow escapeChar to escape quotes even outside quoted fields, it will break previous behavior of ML.NET (of not escaping double quotes) and it would actually be unnecessary (since escaping them was never necessary).

2. **Escaping separators (a.k.a delimitation characters) (outside and/or inside quoted fields)**. Justin [mentioned on his ask](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425735487) that other libraries allow escaping separators, but actually it seems that Databricks only allows [escaping quotes](https://docs.databricks.com/data/data-sources/read-csv.html), whereas Pandas' documentation is more ambiguous, and it says escapechar is used to ["escape other characters"](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)... so I'd still need to actually test if pandas allow escaping separators. Justin has also [shared with me this paper ](https://gertjanvandenburg.com/papers/VandenBurg_Nazabal_Sutton_-_Wrangling_Messy_CSV_Files_by_Detecting_Row_and_Type_Patterns_2019.pdf) where they used escapechar to escape both separators and quotes, but I haven't read it fully yet, and I'm not sure if they conclude anything specifically on the usefulness of escaping separators. Even more, I'm not sure if @justinormont considers this to be a must. Also, it would be ambiguous what to do if the `escapeChar = "` (default) and we encounter the common case of `... "this is a field",5 ...`... should ", be interpreted as an escaped `,`? this tricky case would require more changes to the code, making it somewhat bigger and perhaps somewhat less efficient.

My opinion is we shouldn't allow those two cases, and that we should stick with only having escapeChar escape quotes inside quoted fields, as they might require more code to maintain and have some impact on efficiency. But I'm eager to hear opinions from others. @harishsk .

Reply:
I agree with the above assessment. 

---
In reply to: [427946570](https://github.com/dotnet/machinelearning/pull/5147#discussion_r427946570) [](ancestors = 427946570)

Reply:
I'd recommend covering the first case. 

One possible way to handle it is to only accept quotes when pared w/ a separator:
* Open quote with one of
  * Preceded w/ a separator: `a,"open...close",c` <br/>(with whitespace optional `a, "open...close", c`)
  * At the start of a line: `"open...close",b,c`
* Close quote with one of
  * Followed by a separator: `a,"open...close",c` <br/>(no white space allowed)
  * At the end of a line: `a,b,"open...close"`

```c#
bool isOpenQuote, isCloseQuote, nextCharEscaped;
for (i = 0; i < str.length; i++)
{
  isOpenQuote = !isInsideQuotedRegion && !escapedChar && (
      i == 0 && str[i] == quoteChar
      || i >= 1 && str[i] == quoteChar && str[i-1] == sepChar
      || i >= 2 && str[i] == quoteChar && str[i-2] == sepChar && IsWhiteSpace(str[i-1])
    );

  if (isOpenQuote) { isInsideQuotedRegion = true; }

  isCloseQuote = isInsideQuotedRegion && !isOpenQuote && !escapedChar && (
      str[i] == quoteChar && str[i+1] == sepChar &&  i <= str.length - 2
      || str[i] == quoteChar && i == str.length - 1
    ); 

  if (isCloseQuote) { isInsideQuotedRegion = false; }

  fieldClosed = !isInsideQuotedRegion && (!escapedChar && str[i] == sepChar || i ==  str.length - 1);

  /* if fieldClosed or escapedChar, append to string builder, checking for isOpenQuote/isCloseQuote to skip the start/current char */

  escapedChar = !escapedChar && str[i] == escapeChar && IsEscapable(str[i+1]) && i <= str.length - 2;
}
```

Along with handling the cases of `a,12",b`, `this " quote was escaped`, and `1,these are 2 quotes "" and this is one " quote,true`, this also catches the weird though common enough case of unescaped quotes `a,"open...unescaped"...close",c`.

It wouldn't handle `1,"abc",def"` as the middle quote would still need to be properly escaped.

Reply:
Thanks for taking the time to review this, @justinormont .

In general it would look to me that your logic is correct, and it might be a good (if not the best) to actually fully accept the first case I mentioned above. Although problems would arise if we also wanted to support all of the other current ML.NET behavior... for example, that we currently ignore all spaces before the opening quote and after the closing quote, whereas your logic only accepts 1 space before and none after (and this characteristic is what your logic rely on for distinguishing between a "real" closing quote that's supposed to be used to close the field, or a simple quote that just happens to be inside a field... so we would either have to choose between your logic or ML.NET current behavior). Also, with your logic I'm not sure if the whole QuotingError mechanism in ML.NET (where quoting errors are logged and empty fields are returned if they appear) would make sense.

In any case, as I mention below in [another comment](https://github.com/dotnet/machinelearning/pull/5147#discussion_r428393614) the code of the Parser already avoids as much as possible doing comparisons and calling functions. So it might not be good to add all of the checks that your logic applies to every character of the input string. Because of this, I wouldn't recommend going on with your logic.

Specially, since all the valid cases (according to the RFC) are already accepted, and your logic seems to only be accepting some unescaped quotes (for which we would typically simply have a QuotingError), and being able to escape "escapable characters" inside or outside of quoted fields (which might or might not be the desired behavior of any given user, if we hard code which are these escapable characters).

Without your logic, particularly for case 1 you mentioned might be important, the behavior on this PR is to leave unquoted fields untouched... so `"1,this is an \"unquoted\" field,true"` will read the second field as `this is an \"unquoted\" field` (i.e. it won't escape the `\"` as your logic does, but it won't throw errors or anything else, it will simply not remove the `\`).

So I don't think it's worth it to add more logic to this PR to support escaping in case 1. If that's ok with you.

Reply:
Sounds good. Thanks for thinking through the impact.